apis/common/v1/connection_details.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import "k8s.io/apimachinery/pkg/runtime"
+
+// PublishConnectionDetailsTo represents configuration of a connection secret.
+type PublishConnectionDetailsTo struct {
+	// Name is the name of the connection secret
+	Name string `json:"name"`
+
+	// Metadata is secret store specific key/value pairs to be used as metadata
+	// Please note, expected keys will differ for each store type. For example,
+	// it could be "labels" and "annotations" in case of "Kubernetes", but it
+	// would be "tags" for "AWS Secret Manager".
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Metadata runtime.RawExtension `json:"metadata,omitempty"`
```

Comment:
I'm a little wary of this being unschemafied. There's no chance we could just make it a simple `map[string]string` and then be opinionated about whether the map corresponds to labels, annotations, tags, etc on a case-by-case basis depending on type of secret store, is there?

(Apologies if we already discussed this during design.)

Reply:
That is actually something that I was planning to discuss/change in the [follow-up PR](https://github.com/crossplane/crossplane-runtime/pull/322) adding `Vault` as a secret store. 

The original plan was to have this field as `map[string]interface{}` and each Secret Store implementation could expect to find their own metadata like:

For Kubernetes:

```
spec:
  publishConnectionDetailsTo:
    name: my-db-connection
    metadata:
      labels:
        environment: production
      annotations:
        acme.example.io/secret-type: infrastructure
    configRef:
      name: default
```

For AWS Secrets Manager:

```
spec:
  publishConnectionDetailsTo:
    name: my-db-connection
    metadata:
      tags:
        environment: production
    configRef:
      name: aws-secret-manager-prod
```

I would like to also support configuring `Type` of the secret in the case of Kubernetes as requested [here](https://github.com/crossplane/crossplane/pull/2494#discussion_r754145633).

However, while working on Vault implementation, I noticed that having different metadata per Secret Store would not be convenient. In other words, I would need to change my MR spec (i.e. `publishConnectionDetailsTo`) and even compositions etc. when I want to store to a different secret store.


My current thinking is to have this field as a superset of all and reuse the same fields whenever it makes sense. I mean:

```go
type PublishConnectionDetailsTo struct {
	Metadata ConnectionSecretMetadata `json:"metadata,omitempty"`
	...
}
type ConnectionSecretMetadata struct {
	Labels      map[string]string `json:"labels"`
	Annotations map[string]string `json:"annotations"`
	Type        corev1.SecretType `json:"type"`
}
```

Kubernetes uses all metadata.
Vault uses Labels to set `custom_metadata`
AWS SM uses Labels to set `tags`
and so on...

If some secret store Foo needs a special metadata that does not fit to existing fields we may consider extending common `ConnectionSecretMetadata` type.

What do you think?

Reply:
Just pushed a commit that defines the scheme proposed above to the secret metadata.